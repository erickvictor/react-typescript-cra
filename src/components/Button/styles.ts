import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from '.'

type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'color'>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3.46rem;
    font-size: ${theme.font.sizes.xsmall};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4.48rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.small};
  `,
  withIcon: (theme: DefaultTheme) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    svg {
      width: 2.2rem;

      & + span {
        margin-left: ${theme.spacings.micro};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, color, hasIcon }) => css`
    color: ${theme.colors.white};
    background: ${theme.colors[color!]};
    border: 0;
    cursor: pointer;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    font-family: ${theme.font.family};
    font-weight: ${theme.font.bold};

    ${!!size && wrapperModifiers[size](theme)};
    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
  `}
`
