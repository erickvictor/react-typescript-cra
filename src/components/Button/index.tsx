import React from 'react'
import * as S from './styles'

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'small' | 'medium'
  color?: 'blackGdt' | 'primaryGdt'
  icon?: JSX.Element
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = ({
  children,
  icon,
  color = 'blackGdt',
  size = 'medium',
  ...props
}: ButtonProps) => (
  <S.Wrapper size={size} color={color} hasIcon={!!icon} {...props}>
    {icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
