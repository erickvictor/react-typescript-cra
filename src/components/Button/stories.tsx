import { Story, Meta } from '@storybook/react/types-6-0'
import { ChevronDown } from '@styled-icons/ionicons-solid/ChevronDown'
import Button from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story = (args) => <Button {...args} />

Default.args = {
  children: 'mais populares'
}

export const withIcon: Story = (args) => <Button {...args} />

withIcon.args = {
  children: 'por gênero',
  icon: <ChevronDown />
}
