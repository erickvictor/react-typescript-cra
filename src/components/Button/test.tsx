import React from 'react'
import { screen } from '@testing-library/react'
import { renderWithTheme } from '../../utils/tests/helpers'
import { ChevronDown } from '@styled-icons/ionicons-solid/ChevronDown'
import 'jest-styled-components'

import Button from '.'

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>mais populares</Button>)

    expect(screen.getByRole('button', { name: /mais populares/i })).toHaveStyle(
      {
        height: '4.48rem',
        padding: '0.8rem 2.4rem',
        'font-size': '1.4rem'
      }
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">mais populares</Button>)

    expect(screen.getByRole('button', { name: /mais populares/i })).toHaveStyle(
      {
        height: '3.46rem',
        padding: '0.8rem 1.6rem',
        'font-size': '1.112rem'
      }
    )
  })

  it('should render a black background by default', () => {
    renderWithTheme(<Button>mais populares</Button>)

    expect(
      screen.getByRole('button', { name: /mais populares/i })
    ).toHaveStyleRule(
      'background',
      'linear-gradient(180deg,#212125 0%,#2e2e35 100%)'
    )
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<ChevronDown data-testid="icon" />}>mais populares</Button>
    )

    expect(screen.getByText(/mais populares/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
})
