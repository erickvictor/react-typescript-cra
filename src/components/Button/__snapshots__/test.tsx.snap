// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  color: #EAEAEA;
  background: linear-gradient(180deg,#212125 0%,#2e2e35 100%);
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 600;
  height: 4.48rem;
  font-size: 1.4rem;
  padding: 0.8rem 2.4rem;
}

<button
  class="c0"
  color="blackGdt"
>
  <span>
    mais populares
  </span>
</button>
`
